ops/pebble.py

```diff
@@ -1976,3 +1972,232 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+
+    def get_response(self):
+        """Return the deserialized JSON object from the multipart "response" field."""
+        return self._response
+
+    def filenames(self):
+        """Return a list of filenames from the "files" parts of the response."""
+        return list(self._files.keys())
+
+    def get_file(self, path, encoding):
+        """Return an open file object containing the data."""
+        mode = 'r' if encoding else 'rb'
+        # We're using text-based file I/O purely for file encoding purposes, not for
+        # newline normalization.  newline='' serves the line endings as-is.
+        newline = '' if encoding else None
+        return open(self._files[path].name, mode, encoding=encoding, newline=newline)
+
+
+def _multipart_parser(
+        marker,
+        handle_header,
+        handle_body,
+        handle_done=None,
+        max_lookahead=0,
+        max_boundary_length=0):
+    r"""Returns a stateful function for parsing mime multipart messages.
+
+    Args:
+        marker: the multipar boundary marker (i.e. in "\r\n--<marker>--\r\n")
+
+        handle_header(data): is called once with the entire contents of a part header
+
+        handle_body(data, done=False): is called incrementally as part body
+        data is fed into the parser - its "done" parameter is set to true when
+        the body is complete.
+
+        max_lookahead: maximum amount of bytes to buffer when searching for a complete header.
+
+    Returns:
+        A function you can feed bytes into by calling repeatedly until finished.
+    """
+    buf = bytearray()
+    pos = 0  # current position in buf
+    done = False  # whether we have found the terminal boundary and are done parsing
+    header_terminator = b'\r\n\r\n'
+
+    # Note: RFC 2046 notes optional "linear whitespace" (e.g. [ \t]+) after the boundary pattern
+    # and the optional "--" suffix.  The boundaries strings can be constructed as follows:
+    #
+    #     boundary = \r\n--<marker>[ \t]+\r\n
+    #     terminal_boundary = \r\n--<marker>--[ \t]+\r\n
+    #
+    # 99 is arbitrarily chosen to represent a max number of linear
+    # whitespace characters to help avoid wrongly writing boundary
+    # characters into a (temporary) file.
+    if not max_boundary_length:
+        max_boundary_length = len(b'\r\n--' + marker + b'--\r\n') + 99
+
+    def feed(data: bytes):
+        nonlocal buf, pos, done
+
+        if done:
+            return
+        buf.extend(data)
+
+        while True:
+            # seek to a boundary if we aren't already on one
+            i, n, done = _next_part_boundary(buf, marker)
+            if i == -1 or done:
+                return  # waiting for more data or terminal boundary reached
+
+            if pos == 0:
+                # parse the part header
+                if max_lookahead and len(buf) - pos > max_lookahead:
+                    raise ProtocolError('header terminator not found')
+                if header_terminator not in buf:
```

Comment:
Probably better to use `buf.find(header_terminator)` and assign it to a variable, to avoid the search twice (once here and once below with `buf.index`).

Reply:
done